### PR TITLE
WASM Runtime refactoring

### PR DESCRIPTION
**Pull request by NikVolf**
message arguments are not copied to wasm memory always anymore, they can be retrieved by externs as other blockchain-related stuff

### CODE REVIEW

File: ethcore/wasm/src/lib.rs [hunk 1]
```diff
@@ -121,15 +125,8 @@ impl vm::Vm for WasmInterpreter {
...
-		let d_ptr = runtime.write_descriptor(
-			call_args::CallArgs::new(
-				params.address,
-				params.sender,
-				params.origin,
-				params.value.value(),
-				params.data.unwrap_or(Vec::with_capacity(0)),
-			)
-		).map_err(|e| Error(e))?;
+		let d_ptr = runtime.write_descriptor(params.data.unwrap_or(Vec::with_capacity(0)))
```
Comment (rphmeier) [inline ethcore/wasm/src/lib.rs]: `Vec::new` is guaranteed not to allocate. `unwrap_or_else` would be nicer looking for the same behavior

File: ethcore/wasm/src/lib.rs [hunk 2]
```diff
@@ -121,15 +125,8 @@ impl vm::Vm for WasmInterpreter {
...
+			.map_err(|e| Error(e))?;
```
Comment (rphmeier) [inline ethcore/wasm/src/lib.rs]: `map_err(Error)`

File: ethcore/wasm/src/runtime.rs [hunk 3]
```diff
@@ -442,10 +434,10 @@ impl<'a, 'b> Runtime<'a, 'b> {
 	}
 
 	/// Write call descriptor to wasm memory
-	pub fn write_descriptor(&mut self, call_args: CallArgs) -> Result<WasmPtr, InterpreterError> {
+	pub fn write_descriptor(&mut self, input: Vec<u8>) -> Result<WasmPtr, InterpreterError> {
```
Comment (rphmeier) [inline ethcore/wasm/src/runtime.rs]: seems to only require an `&[u8]` and not ownership of a `Vec`

File: ethcore/wasm/src/runtime.rs [hunk 4]
```diff
@@ -457,11 +449,7 @@ impl<'a, 'b> Runtime<'a, 'b> {
 		self.memory.set(d_ptr, &d_buf)?;
 
 		// write call args to memory
-		self.memory.set(args_ptr, &call_args.address)?;
-		self.memory.set(args_ptr+20, &call_args.sender)?;
-		self.memory.set(args_ptr+40, &call_args.origin)?;
-		self.memory.set(args_ptr+60, &call_args.value)?;
-		self.memory.set(args_ptr+92, &call_args.data)?;
```
Comment (rphmeier) [inline ethcore/wasm/src/runtime.rs]: now that these are removed, why is the name of this function still accurate?

**Comment by NikVolf** [comment]
@rphmeier updated, thanks

File: ethcore/wasm/src/tests.rs [hunk 5]
```diff
@@ -88,7 +88,7 @@ fn logger() {
 	};
 
 	println!("ext.store: {:?}", ext.store);
-	assert_eq!(gas_left, U256::from(98417));
+	assert_eq!(gas_left, U256::from(98731));
```
Comment (gavofyork) [inline ethcore/wasm/src/tests.rs]: missing `_` for consistency with the others.

File: ethcore/wasm/src/tests.rs [hunk 6]
```diff
@@ -267,15 +269,15 @@ fn create() {
...
+			gas: U256::from(99_144),
 			sender_address: None,
 			receive_address: None,
 			value: Some(1_000_000_000.into()),
 			data: vec![0u8, 2, 4, 8, 16, 32, 64, 128],
 			code_address: None,
 		}
 	));
-	assert_eq!(gas_left, U256::from(98_860));
+	assert_eq!(gas_left, U256::from(99113));
```
Comment (gavofyork) [inline ethcore/wasm/src/tests.rs]: missing `_`

File: ethcore/wasm/src/tests.rs [hunk 7]
```diff
@@ -441,7 +443,7 @@ fn keccak() {
 	};
 
 	assert_eq!(H256::from_slice(&result), H256::from("68371d7e884c168ae2022c82bd837d51837718a7f7dfb7aa3f753074a35e1d87"));
-	assert_eq!(gas_left, U256::from(84003));
+	assert_eq!(gas_left, U256::from(84026));
```
Comment (gavofyork) [inline ethcore/wasm/src/tests.rs]: missing `_` (and others below)

File: ethcore/wasm/src/runtime.rs [hunk 8]
```diff
@@ -559,6 +547,39 @@ impl<'a, 'b> Runtime<'a, 'b> {
 		Ok(None)
 	}
 
+	fn sender(&mut self, context: InterpreterCallerContext)
+		-> Result<Option<interpreter::RuntimeValue>, InterpreterError>
+	{
+		let return_ptr = context.value_stack.pop_as::<i32>()? as u32;
```
Comment (gavofyork) [inline ethcore/wasm/src/runtime.rs]: why doesn't `context.value_stack.pop_as::<u32>()?` work?

**Comment by gavofyork** [comment]
looks ok to me except for the questions/formatting.

File: ethcore/wasm/src/runtime.rs (hunk 8 above)
Comment (NikVolf) [reply ethcore/wasm/src/runtime.rs]: it's not native webassembly type
there are only four `i32`, `i64`, `f32`, `f64`

**Pull request merged**